rust_src/src/buffers.rs

```diff
@@ -871,6 +871,57 @@ pub fn force_mode_line_update(all: LispObject) -> LispObject {
     all
 }
 
+/// Return a list of overlays which is a copy of the overlay list
+/// LIST, but for buffer BUFFER.
+#[no_mangle]
+pub extern "C" fn copy_overlays(
+    buffer: *mut Lisp_Buffer,
+    list: *mut Lisp_Overlay,
+) -> *mut Lisp_Overlay {
+    if list == ptr::null_mut() {
+        return list;
+    }
+
+    let overlays = LispOverlayRef::from_ptr(list as *mut c_void)
+        .unwrap_or_else(|| panic!("Invalid overlay reference."));
+
+    let mut result: Option<LispOverlayRef> = None;
+    let mut tail: Option<LispOverlayRef> = None;
+
+    for overlay in overlays.iter() {
+        let mk_start = overlay.start().as_marker_or_error();
+        let start = build_marker(
+            buffer,
+            mk_start.charpos_or_error(),
+            mk_start.bytepos_or_error(),
+        );
+        start
+            .as_marker_or_error()
+            .set_insertion_type(mk_start.insertion_type());
+
+        let mk_end = overlay.end().as_marker_or_error();
+        let end = build_marker(buffer, mk_end.charpos_or_error(), mk_end.bytepos_or_error());
+        end.as_marker_or_error()
+            .set_insertion_type(mk_end.insertion_type());
+
+        let overlay_new = unsafe { build_overlay(start, end, Fcopy_sequence(overlay.plist())) };
+
+        if result.is_some() {
```

Comment:
Why is this different from the C code? next and tail should be the same righ?

Reply:
Overall this use of Option just feels clunky.

Reply:
> Why is this different from the C code? next and tail should be the same righ?

Type of tail and next are different here.

> Overall this use of Option just feels clunky.

True, `C` null checks `tail` to make decisions. I couldn't find a way other than `Option` to that.

Reply:
I will put some thought to it. This is likely a pattern we need to repeat. We might as well get it right now.